templates/democracy/contracts/DemocracyTemplate.sol

```diff
@@ -0,0 +1,90 @@
+pragma solidity 0.4.24;
+
+import "@aragon/templates-shared/contracts/BaseTemplate.sol";
+
+
+contract DemocracyTemplate is BaseTemplate {
+    string private constant ERROR_MISSING_TOKEN_CACHE = "DEMOCRACY_MISSING_TOKEN_CACHE";
+    string private constant ERROR_INVALID_HOLDERS_STAKES_LEN = "DEMOCRACY_INVALID_HOLDERS_STAKES_LEN";
+
+    mapping (address => address) internal tokenCache;
+
+    constructor(DAOFactory _daoFactory, ENS _ens, MiniMeTokenFactory _miniMeFactory, IFIFSResolvingRegistrar _aragonID)
+        BaseTemplate(_daoFactory, _ens, _miniMeFactory, _aragonID)
+        public
+    {
+        // solium-disable-previous-line no-empty-blocks
+    }
+
+    function newTokenAndInstance(
+        string tokenName,
+        string tokenSymbol,
+        string id,
+        address[] holders,
+        uint256[] stakes,
+        uint64 supportNeeded,
+        uint64 minAcceptanceQuorum,
+        uint64 voteDuration
+    )
+        public
+    {
+        newToken(tokenName, tokenSymbol);
+        newInstance(id, holders, stakes, supportNeeded, minAcceptanceQuorum, voteDuration);
+    }
+
+    function newToken(string name, string symbol) public returns (MiniMeToken) {
+        MiniMeToken token = _createToken(name, symbol);
+        _cacheToken(token, msg.sender);
+        return token;
+    }
+
+    function newInstance(string id, address[] holders, uint256[] stakes, uint64 supportNeeded, uint64 minAcceptanceQuorum, uint64 voteDuration)
+        public
+    {
+        require(holders.length == stakes.length, ERROR_INVALID_HOLDERS_STAKES_LEN);
+        MiniMeToken token = _popTokenCache(msg.sender);
+
+        // Create DAO and install apps
+        (Kernel dao, ACL acl) = _createDAO();
+        Vault vault = _installVaultApp(dao);
+        Finance finance = _installFinanceApp(dao, vault, 30 days);
+        TokenManager tokenManager = _installTokenManagerApp(dao, token, true, uint256(-1));
+        Voting voting = _installVotingApp(dao, token, supportNeeded, minAcceptanceQuorum, voteDuration);
+
+        // Mint tokens
+        _createPermission(acl, tokenManager, tokenManager.MINT_ROLE());
+        for (uint256 i = 0; i < holders.length; i++) {
+            tokenManager.mint(holders[i], stakes[i]);
+        }
+        _removePermission(acl, tokenManager, tokenManager.MINT_ROLE());
+
+        // Set up permissions
+        _createVaultPermissions(acl, vault, finance, voting);
+        _createFinancePermissions(acl, finance, voting, voting);
+        _createTokenManagerPermissions(acl, tokenManager, voting, voting);
+        _createEvmScriptsRegistryPermissions(acl, voting, voting);
+        _createVotingPermissions(acl, voting, tokenManager);
+        _transferPermission(acl, voting, dao, dao.APP_MANAGER_ROLE());
+        _transferPermission(acl, voting, acl, acl.CREATE_PERMISSIONS_ROLE());
+
+        _registerID(id, dao);
+    }
+
+    function _createVotingPermissions(ACL acl, Voting voting, TokenManager tokenManager) internal {
```

Comment:
This overloading is a bit scary, what do you think about naming it something else?

Reply:
Hm, scary why? What could happen? I actually kind of like it, as the base one is missing `CREATE_VOTES_ROLE`. Plus the signature is different.

Reply:
> What could happen?

You think you're using the base one, but actually not. Or more likely, somewhere else down the line, someone thinks they're using the base one, but they get this version instead.

Reply:
> You think you're using the base one, but actually not. 

But that's always a problem with overloading, right? Besides here the signature is different, so it makes it less likely.

> I find it primarily scary here because it's not intended to be overloaded; it's not really extensible functionality in my eyes.

As I said before, I disagree with this, because permissions setting in the first one is incomplete, so it makes total sense to extend it with the ones missing.

But of course perfectly fine for me if you prefer to rename it.

Reply:
In other templates I used the `_createCustom...Permissions`, will apply that to all the templates